File: leetcode/easy/BalancedBInaryTree.py
from typing import Optional

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        """
        Determines if a binary tree is height-balanced.

        A height-balanced binary tree is a binary tree in which the depth of the two
        subtrees of every node never differs by more than one.

        This solution uses a recursive, bottom-up Depth-First Search (DFS) approach.
        The helper function `dfs` serves a dual purpose:
        1. It calculates the height of a subtree.
        2. It signals if a subtree is unbalanced by returning a special value (-1).

        This is more efficient than the naive approach because it stops the recursion
        as soon as an imbalance is found.
        """
        def dfs(node: Optional[TreeNode]) -> int:
            """
            Recursively checks the height and balance of a subtree.
            Returns the height of the subtree if it's balanced, otherwise returns -1.
            """
            # Base case: An empty tree is balanced and has a height of 0.
            if not node:
                return 0

            # Recursively get the height of the left subtree.
            left_height = dfs(node.left)
            # If the left subtree is unbalanced, propagate the unbalanced signal (-1) up.
            if left_height == -1:
                return -1

            # Recursively get the height of the right subtree.
            right_height = dfs(node.right)
            # If the right subtree is unbalanced, propagate the signal up.
            if right_height == -1:
                return -1

            # Check if the current node is unbalanced.
            if abs(left_height - right_height) > 1:
                return -1

            # If balanced, return the height of the current subtree.
            return 1 + max(left_height, right_height)

        # The tree is balanced if the final result from the root is not -1.
        return dfs(root) != -1

File: leetcode/easy/BalancedBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    /**
     * Checks if a binary tree is height-balanced.
     * A height-balanced binary tree is a binary tree in which the depth of the
     * two subtrees of every node never differs by more than one.
     *
     * This solution uses a single-pass, post-order traversal (DFS) approach,
     * which is optimal with O(n) time complexity.
     *
     * @param root The root of the binary tree.
     * @return true if the tree is height-balanced, false otherwise.
     */
    public boolean isBalanced(TreeNode root) {
        // The helper function `getHeight` returns the height of the tree if it's balanced,
        // or -1 if it's not. We just need to check the final result.
        return getHeight(root) != -1;
    }

    /**
     * A helper function that performs a post-order traversal to both check for
     * balance and calculate the height of a subtree.
     *
     * This function has a dual purpose:
     * 1. It returns the height of the node if its subtree is balanced.
     * 2. It returns -1 as a special signal if it detects that the subtree is unbalanced.
     *
     * This "early exit" signal (-1) avoids redundant computations that would occur
     * in a naive top-down approach.
     *
     * @param node The root of the subtree to check.
     * @return The height of the subtree if balanced, otherwise -1.
     */
    private int getHeight(TreeNode node) {
        // Base case: An empty tree is balanced and has a height of 0.
        if (node == null) {
            return 0;
        }

        // Recursively find the height of the left subtree.
        // If the left subtree is already unbalanced, propagate the -1 signal up.
        int leftHeight = getHeight(node.left);
        if (leftHeight == -1) {
            return -1;
        }

        // Recursively find the height of the right subtree.
        // If the right subtree is already unbalanced, propagate the -1 signal up.
        int rightHeight = getHeight(node.right);
        if (rightHeight == -1) {
            return -1;
        }

        // Check if the current node is balanced.
        // The difference in heights of its subtrees must not be more than 1.
        if (Math.abs(leftHeight - rightHeight) > 1) {
            return -1; // Signal that the tree is unbalanced at this node.
        }

        // If the current node is balanced, return its height.
        // The height of a node is 1 + the height of its tallest subtree.
        return 1 + Math.max(leftHeight, rightHeight);
    }
}
